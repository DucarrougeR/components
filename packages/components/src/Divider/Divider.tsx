import {
  CompatibleHTMLProps,
  color,
  position,
  PositionProps,
  reset,
  space,
  SpaceProps,
} from '@looker/design-tokens'
import styled from 'styled-components'
import { variant } from 'styled-system'

export interface DividerProps
  extends CompatibleHTMLProps<HTMLHRElement>,
    PositionProps,
    SpaceProps {
  size?: string | number
  customColor?: string
  appearance?: 'default' | 'light' | 'dark' | 'onDark'
}

const appearanceVariant = variant({
  prop: 'appearance',
  variants: {
    dark: {
      bg: 'palette.charcoal400',
    },
    default: {
      bg: 'palette.charcoal300',
    },
    light: {
      bg: 'palette.charcoal200',
    },
    onDark: {
      bg: 'palette.charcoal500',
    },
  },
})

export const Divider = styled.hr.attrs((props: DividerProps) => ({
  bg: props.customColor,
}))<DividerProps>`
  ${reset}
  ${position}
  ${space}

  height: ${props => props.size};
  ${props => (props.customColor ? color : appearanceVariant)}
`

Divider.defaultProps = { appearance: 'default', size: '1px' }
