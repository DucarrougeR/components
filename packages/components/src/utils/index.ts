export * from './GlobalStyle'
export * from './useToggle'
