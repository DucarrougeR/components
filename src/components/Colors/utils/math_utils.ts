/**
 * Utility to convert a numerical value to a percent given the value and a max value.
 */
export const toPerc = (value: number, maxValue: number) =>
  Math.round((value / maxValue) * 100)
