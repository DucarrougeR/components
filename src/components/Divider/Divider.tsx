import * as React from 'react'
import { css, styled } from '../../style'
import { ThemedProps } from '../../types'
import { Box, BoxPropsWithout } from '../Box'

interface DividerProps extends BoxPropsWithout<HTMLDivElement, 'color'> {
  size?: string | number
  customColor?: string
  variant?: 'light' | 'dark' | 'onDark'
}

function dividerVariant(props: ThemedProps<DividerProps>) {
  switch (props.variant) {
    case 'light':
      return css`
        background: ${props.theme.colors.palette.charcoal200};
      `
    case 'dark':
      return css`
        background: ${props.theme.colors.palette.charcoal400};
      `
    case 'onDark':
      return css`
        background: ${props.theme.colors.palette.charcoal500};
      `
    default:
      return false
  }
}

const InternalDivider: React.SFC<DividerProps> = ({
  size,
  customColor,
  variant,
  ...props
}) => {
  return (
    <Box
      height={size || '1px'}
      bg={customColor || 'palette.charcoal300'}
      {...props}
    />
  )
}

export const Divider = styled<DividerProps>(InternalDivider)`
  ${dividerVariant};
`
