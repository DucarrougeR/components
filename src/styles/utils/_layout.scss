// Returns a value from the $spacing list
@function spacing($key) {
  @return map-lookup($spacing, $key, '$spacing');
}
