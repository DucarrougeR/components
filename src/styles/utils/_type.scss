// Returns a value from the $test-size list
@function font-size($key) {
  @return map-lookup($text-size, $key, '$text-size');
 }

 // Returns a value from the $line-height list
 @function line-height($key) {
   @return map-lookup($line-height, $key, '$line-height');
 }

 @mixin fontSize($size) {
  font-size: $size;
  font-size: pxToRem($size);
}
