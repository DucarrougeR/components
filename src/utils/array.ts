/**
 * Returns an array of length `size` filled with zeros.
 * Useful for looping constructs where a length is known, but no array is provided.
 */

export const sizedArray = (size: number) => Array(size).fill(0)
