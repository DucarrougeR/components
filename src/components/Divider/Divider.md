```js noeditor
const StatusAndResources = require('../../../styleguide_components/StatusAndResources')
  .StatusAndResources

;<StatusAndResources
  status="stable"
  figmaURL="https://www.figma.com/file/FJdRcUtLJ6RVIcPS1CL5dwwQ/Borders-and-Dividers?node-id=7%3A10"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Divider/Divider.tsx"
  feedbackTitle="Divider Component Feedback"
/>
```

## Default Divider

Dividers accept all the spacing props from `<Box />`, which gives you the ability to easily put spacing around your dividers. By default a divider's background color will be `charcoal300`

```js
import { Divider } from './Divider'
import { Box } from '../Box'
import { Flex } from '../Flex'
;<Flex justifyContent="space-around">
  <Box bg="white " p="xlarge" mr="large">
    On White
    <Divider mt="medium" />
  </Box>
  <Box bg="palette.charcoal000" p="xlarge" mr="large">
    On Charcoal000
    <Divider mt="medium" />
  </Box>

  <Box bg="palette.charcoal100" p="xlarge">
    On Charcoal100
    <Divider mt="medium" />
  </Box>
</Flex>
```

## Divider Appearance

There are three divider appearances `light`, `dark` and `onDark`.

```js
import { Divider } from './Divider'
import { Box } from '../Box'
import { Flex } from '../Flex'
;<Flex justifyContent="space-around">
  <Box bg="white " p="xlarge" mr="large">
    light appearance
    <Divider mt="medium" appearance="light" />
  </Box>
  <Box bg="palette.charcoal000" p="xlarge" mr="large">
    dark appearance
    <Divider mt="medium" appearance="dark" />
  </Box>

  <Box bg="palette.charcoal600" p="xlarge">
    onDark appearance
    <Divider mt="medium" appearance="onDark" />
  </Box>
</Flex>
```

## Custom Divider

You can adjust the dividers height and supply a custom color if needed

```js
import { Divider } from './Divider'
import { Box } from '../Box'
;<Box p="xlarge">
  Custom Divider
  <Divider size="8px" customColor="turquoise" mt="large" borderRadius="100px" />
</Box>
```
