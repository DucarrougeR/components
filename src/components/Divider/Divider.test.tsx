import 'jest-styled-components'
import * as React from 'react'
import { assertSnapshot } from '../../../test/utils/snapshot'
import { Divider } from './Divider'

test('Default Divider', () => {
  assertSnapshot(<Divider />)
})

test('Divider with a variant', () => {
  assertSnapshot(<Divider variant="onDark" />)
})

test('A custom Divider', () => {
  assertSnapshot(<Divider size="20px" customColor="turquoise" />)
})
