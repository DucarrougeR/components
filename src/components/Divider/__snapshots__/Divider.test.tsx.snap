// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A custom Divider 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  background-color: turquoise;
  height: 20px;
}

<div
  className="c0"
/>
`;

exports[`Default Divider 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  background-color: #C1C6CC;
  height: 1px;
}

<div
  className="c0"
/>
`;

exports[`Divider with a variant 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  background-color: #C1C6CC;
  height: 1px;
}

.c0 {
  background: #707781;
}

<div
  className="c0 c1"
/>
`;
